rfcs/0151-nixos-port-alloc.md

```diff
@@ -0,0 +1,135 @@
+---
+feature: nixos-port-alloc
+start-date: 2023-06-10
+author: lucasew
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+A port allocator for NixOS services.
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes people don't care about which port a service is running, only that it
+should be written somewhere so a service such as nginx can find it.
+
+# Detailed design
+[design]: #detailed-design
+
+The less likely used ports of the port space in a system are the higher ones,
+and the highest one is 65535, so a NixOS module could keep track of which services
+need a port and the service modules need only to reference that port in their
+configurations.
+
+This module exposes the options under `networking.ports`. A service module can
```

Comment:
Considering the wording in here, this is supposed to be opt-in, correct? I.e. this won't be used (or enforced to be used) in other upstream NixOS modules itself, correct?

Reply:
Yes. The allocator is opt-in. I specially use for services that are behind nginx so I don't need to think about which port is being used by each service (I got a few weird errors and conflicts before that and wanted to keep ports such as 3000 and 5000 free).

Reply:
I wonder if it makes sense to use it by default, however in this case it would probably need to support a "preferred port". I'm thinking something like this:

1. Services set a "preferred port" and register with the allocator.
2. Upon collision all but one of the colliding services are reassigned.
3. There is an option to force a specific port for a service (and get an eval error if there is a conflict).

So the options would be something like:

- `Prefer(port)`
- `Require(port)`
- `NoPreference`

The downside would be that if you enable a new service it may shift an existing service which could cause issues if you were relying on the default. (For example you were running Mumble and connecting to mumble://yourhost.example without a port) This could cause runtime issues after switching to the new configuration. This is probably worse than an switch-time failure to start the new service. However for the reverse-proxy case it makes a lot of sense to have this behaviour as you can configure your reverse-proxy using something like `config.service.foo.port` and it would auto-update.

In fact maybe it would make sense to use the port allocator for all services by default but use `Require(port)`. That way 1. The port allocator won't assign that port to another service and 2. You will get an eval-time error if you enable two services that use the same port. Then the user could configure an override with a different port or with `NoPreference`. New services that bind to an internal IP could maybe considering using `NoPreference` by default as it won't cause any backwards incompatibility.

Reply:
If this should be actually used by default, another thing to check whether (and how much) this will increase the expensiveness of evaluation a NixOS system.